docs/architecture.md

```diff
@@ -0,0 +1,176 @@
+* The ResponseCache -- essentially just the UnverifiedBlockStore + some link tracking, so that the BlockStore is properly pruned as requests expire
+* The LoadAttemptQueue -- Basically a queue to track attempts to load links. A link load can have one of three results --
+- It can load successfully,
+- It can load and knowingly fail (i.e responder tells you it is for sure missing a link)
+- It can be indeterminate -- you don't yet have the block in the response cache or in local storage, but the responder is still sending responses and hasn't indicated it doesn't have the block
+The load attempt queue essentially attempts the load then if the response is 3, puts the load attempt in a paused queue awaiting more responses (note if the responder eventually finishes without ever sending a response for that link, the request will error at the point the last response is sent)
+
+With the AsyncLoader handling the complicated parts of loading data from the network, the RequestManager simply manages the overall process.
+
+## Responder Implementation
+
+To respond to a request, the responder implementation needs to:
+
+* Decode the selector
+* Initiate a selector traversal with go-ipld-prime
+* Provide ipld-prime with an overloaded link loader
+* Whenever IPLD prime loads a link during traversal, use the intercepted loader to send a message across the network to the requestor with the block and/or metadata about the block
+* Terminate the response
+
+In addition, an optimized responder implementation accounts for the following concerns:
+
+* *"Don't get DDOS'd"* - a denial of service attack should not be trivially easy. Selector traversal carries a non-trivial CPU and memory cost, so the responder needs to take care not to simply executes every graphsync query it receives immediately.
+
+* *Preserve Bandwith* - Be efficient with network usage, dedepulicate data, and buffer response output so that each new network message contains all response data we have at the time the pipe becomes free.
+
+The responder implementation is managed by the Response Manager. The ResponseManager delegates to PeerTaskQueue to rate limits the number of of in progress selector traversals and insure no one peer is given more priority than others. As data is generated from selector traversals, the ResponseManager uses the PeerResponseManager to aggregate response data for each peer and send compact messages over the network.
+
+The follow diagram outlines in greater detail go-graphsync's responder implementation, covering how it's initialized and how it responds to requests:
+![Responding To A Request](responder-sequence.png)
+
+Here are some key components in this implementation
+
+### PeerTaskQueue - Preventing DOS Attacks
+
+Rather than responding to incoming requests immediately, the ResponseManager places each incoming request is placed on a prioritized queue. 
```

Comment:
```suggestion
Rather than responding to incoming requests immediately, the ResponseManager places each incoming request in a prioritized queue. 
```